### DISCUSSION TITLE
Heavy traffic from Renovate on the Gradle Plugin Portal

### DISCUSSION THREAD

**Discussion by ljacomet** [renovatebot/renovate#27018] [Request Help]
### How are you running Renovate?

None

### If you're self-hosting Renovate, select which platform you are using.

None

### What is your question?

Hey Renovate team,

We at Gradle are offering the [Gradle Plugin Portal](https://plugins.gradle.org/) for Gradle builds to download Gradle plugins, mostly community contributed.

Over the past 2 weeks we have seen the Renovate traffic increase a lot against the Gradle Plugin Portal, to the point where it impacts [its availability](https://status.gradle.com/incidents/2c40qrs5pm8y).

<img width="793" alt="20240202_RenovateTraffic" src="https://github.com/renovatebot/renovate/assets/135308/6371d244-cbb4-4816-84fe-51f32cb28a3b">

We have identified a single IP as causing most of the added traffic and have taken measures for this. But even without it, the traffic almost doubled over a month.

Are you aware of recent changes in Renovate that could have caused increased traffic related to Gradle plugins in Gradle projects?
Or have you registered a similar spike in Renovate usage, in particular on Gradle projects?

Regards,
Louis Jacomet
Gradle Build Tool team

### Logs (if relevant)

<details><summary>Logs</summary>

```

Copy/paste the relevant log(s) here, between the starting and ending backticks

```

</details>

**Comment by rarkins** [comment]
Hi Louis, we'll look into it ASAP. Most likely scenario is a regression in caching of lookups

**Comment by rarkins** [comment]
https://github.com/renovatebot/renovate/discussions/27019 could be the same problem. 37.139.4 was released on January 17: https://github.com/renovatebot/renovate/releases/tag/37.139.4. The graph above shows a bit of a ramp up from there and then a bigger rise starting around 28th January.

If most was from one IP, it's probably the Mend hosted app. @nabeelsaabna on which day did we upgrade from <=37.139.3 to >=37.139.4?

Cc @zharinov

**Reply by ljacomet** [reply]
Thanks for the pointers. "constant cache miss" looks like a good reason why we would see such an increase in requests. 😅 

Let me know if you need more specific information from traffic on our end. I could share the details about that specific IP by email if needed.

**Reply by rarkins** [reply]
Fix PR for identified cache problem: https://github.com/renovatebot/renovate/pull/27021

Strange thing is it was released 2 weeks ago but we got the first two reports about problems today within an hour of each other

@ljacomet very sorry for impacting your portal!

**Reply by ljacomet** [reply]
We identified that the issue was Renovate related yesterday, Wed Feb 1 '24, but did not immediately think about reaching out. That one is on us.
Glad that this is going to be fixed and that our traffic should return to normal soon.

Can you update here when the Mend app is updated so we can see if it has a direct effect?

**Reply by rarkins** [reply]
Should be fixed in https://github.com/renovatebot/renovate/releases/tag/37.168.2

Docker images will be built soon.

This means that within a few hours you should see self-hosted Renovate users start to decrease their load on your portal as they adopt the new version. I would expect the decrease to be similar to the slow increase you saw after January 17 (it would have been caused by self-hosted users updating their Renovate to latest and using it, now when they update to latest they should have a fixed behavior which will be optimized as soon as the cache is written). The Mend hosted app (which I assume is the IP you blocked) should be updated within 48 hours maximum, and we will confirm here when done

**Reply by rarkins** [reply]
@ljacomet can you confirm that traffic has returned to normal and that you’re now able to unblock our AWS IP? Also if you know who filed the abuse report to AWS can they withdraw/resolve it so that AWS doesn’t take down the entire Renovate app?

**Reply by ljacomet** [reply]
I can confirm traffic is back to normal. All good on our end. Already discussed withdrawing the abuse report. Will confirm once done.

**Reply by aaltounyan** [reply]
Hi @rarkins, traffic from the EC2 IP ending in .110 has now settled back at it's previous normal level. At it's peak around Feb 2nd it was running 1500 requests per second and now back to an average of 25/s, though it still occasionally peaks above the rate limit. 

We never blocked any IPs but enforced a specific rate limit of around 100/s and returned a 429 to anything above that limit. Does Renovate handle 429s gracefully? We are considering strengthening the general rate limit to protect the availability of the Plugin Portal. We will announce this publicly if we roll it out.

Finally, yes i will get back in touch with the AWS Abuse team and let them know. We had no way to identify the owner of this IP and needed to protect our availability and DoS protection bills.

**Reply by viceice** [reply]
yes, renovate should handle 429 gracefully

**Reply by rarkins** [reply]
@aaltounyan thanks for the info, we understand completely and sorry again for the bug!

**Reply by aaltounyan** [reply]
Hi team

We are having to update the way our rate limit functions due to imposed changes by our provider. We expect this change to happen imminently this week. 

Until now, only requests above the limit (100/s) were blocked, but going forward once the limit is reached, subsequent requests will be blocked with a 429 for 30 seconds. I'm not sure how this will affect your Mend hosted app, but wanted to give you a heads up. 

Your .110 IP is still occasionally hitting the limit, with 2.6K 429s out of 2.2M requests in the last 24 hours, but this hasn't been causing any problems for us. 

We are working with our provider on this, and at the moment cannot say for sure if this rule will stay as-is or return to the previous state. Please let us know if you are impacted by this change.

**Reply by HonkingGoose** [reply]
### Thundering herd problem?

> Until now, only requests above the limit (100/s) were blocked, but going forward once the limit is reached, subsequent requests will be blocked with a 429 for 30 seconds. I'm not sure how this will affect your Mend hosted app, but wanted to give you a heads up.

Thank you for the heads-up! ❤️ Checking if I understand you correctly:

1. A Renovate client (ours, or self-hosted user) overwhelms your server.
2. Your server returns a `429` reponse with 30 seconds wait time.
3. Client waits 30 seconds, and then fires a new request at your server.

So far, so good, but what if:

4. All clients come back at the _exact_ same time, thus still overwhelming your server, just 30 seconds _later_? [Wikipedia: Thundering herd problem](https://en.wikipedia.org/wiki/Thundering_herd_problem)

#### Possible solutions to thundering herd

- Buckets with range of return times
- Exponential backoff
- Setting hard limits for Renovate clients in our code

##### Buckets with a range of return times

Would it be a good idea to add a bit of randomness to the `429` wait times your server hands out? Like one bucket of clients waits between 30-35 seconds, then the next bucket 35-40, the next 40-45, and so on. That way you break up the load into smaller chunks?

##### Exponential backoff

Or maybe you can use exponential backoff, where if your server is still overloaded, you return progressively longer wait times?

##### Set hard limits for Renovate clients

Maybe we should set hard limits for the number of requests Renovate is allowed to fire at your server? I asked the maintainers if that's a good idea, see:

- https://github.com/renovatebot/renovate/issues/27802

### Summary

I listed some possible solutions, and ideas. I'll let the Renovate maintainers, and you, decide what's best. 😉

**Reply by aaltounyan** [reply]
It's per IP, so general users of Renovate won't be impacted. We know from the history of this Issue that the hosted Mend service may be impacted. The servers shouldn't be impacted, our DDoS provider would just block them again. From what we can see, there are only occasional peaks where Mend goes over the limit - the workloads behind those peaks may suffer.

We are limited in our options based on our provider, so just explaining how things are likely to work for the time being. The other options you mention are not possible at this moment.

**Comment by jamietanna** [comment]
Closing as hopefully this should be resolved - @aaltounyan thanks for flagging, and please re-open/raise a fresh note if any further action is needed!
